### ISSUE TITLE
Exception thrown when cancelling

### ISSUE THREAD

**Issue by mikasR** [paypal/android-checkout-sdk#6]
@CodyEngel Hello, i'm having another issue.

If i tap on the cancel label my app crashes with the following exception:

Fatal Exception: kotlin.UninitializedPropertyAccessException: lateinit property accessToken has not been initialized
       at com.paypal.pyplcheckout.utils.PYPLCheckoutUtils.getCheckoutLiteUrl(PYPLCheckoutUtils.kt:235)
       at com.paypal.pyplcheckout.utils.PYPLCheckoutUtils.getCheckoutLiteUrl$default(PYPLCheckoutUtils.kt:167)
       at com.paypal.pyplcheckout.utils.PYPLCheckoutUtils.fallBackToWeb(PYPLCheckoutUtils.kt:335)
       at com.paypal.pyplcheckout.utils.PYPLCheckoutUtils.fallBack(PYPLCheckoutUtils.kt:309)
       at com.paypal.pyplcheckout.home.viewmodel.MainPaysheetViewModel$3.authFailure(MainPaysheetViewModel.java:537)
       at com.paypal.pyplcheckout.flavorauth.ThirdPartyAuth$2.completeWithFailure(ThirdPartyAuth.java:122)
       at com.paypal.authcore.authentication.Authenticator$a.onReceive(:75)
       at androidx.localbroadcastmanager.content.LocalBroadcastManager.executePendingBroadcasts(LocalBroadcastManager.java:313)
       at androidx.localbroadcastmanager.content.LocalBroadcastManager$1.handleMessage(LocalBroadcastManager.java:121)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:223)
       at android.app.ActivityThread.main(ActivityThread.java:7656)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

**Comment by CodyEngel** [comment]
Hey thanks @mikasR, I haven't seen this crash through testing 🤔 From looking at the code path it looks like you were in a web fallback flow and the stacktrace definitely helps out quite a bit.

When you get a chance could you attach a screenshot of what you see just before the app crash as well as the exact steps you took to reproduce? If you are using a payment button can you also specify which one you were using?

Thanks, in the meantime I'll do some initial investigation into how this might be happening.

**Comment by mikasR** [comment]
@CodyEngel pretty much nothing happens. I tap cancel and the screen goes black, and the app restarts.

I am using a paymentbutton.PayPalButton.
All i did was setup the button. Tap on it, and on the paypal webview i tap cancel the behaviour is as described above.

![image](https://user-images.githubusercontent.com/10234099/115903375-69dcf680-a45b-11eb-979a-fa7b8b1885bf.png)

**Comment by mikasR** [comment]
Also, on another note.
If i login and accept the transaction, i get an onError event with the following:
ErrorInfo(error=com.paypal.pyplcheckout.exception.PYPLException: exception when creating order: exception when creating order: 400, reason=Error creating order, correlationIds=CorrelationIds(eligibilityDebugID=null, fundingEligibilityDebugID=null, updateClientConfigDebugID=6e1e07473311, lsatUpgradeDebugID=null, fetchPayloadDebugID=null, currencyConversionDebugID=null, finishCheckoutDebugID=null), orderId=null, nativeSdkVersion=1.8.5)

**Comment by CodyEngel** [comment]
Awesome, thank you @mikasR I will take a look and will keep you updated.

**Comment by mikasR** [comment]
@CodyEngel This last exception is solved.
Only remains the one after cancellation.

**Comment by CodyEngel** [comment]
Hey @mikasR -- wanted to give a quick update (our offices were closed on Monday for a wellness day, sorry I should have mentioned this on Friday 😫).

I'm currently working on a fix for the cancellation bug. It looks like that cancellation text was added sometime after we originally implemented authentication, and unfortunately the cancellation reason given when you tap that text as opposed to the "X" in the chrome custom tab differ. At this point I'm wrapping some tests around this.

Once the fix is in place my plan is to release this first as a snapshot if you are comfortable with validating the bug fix. I'll include instructions for adding the snapshot on this issue once it becomes available (hopefully in the next day or two).

**Comment by mikasR** [comment]
Hi @CodyEngel.

Thanks. I can validate the fix with snapshot, no problem.

**Comment by CodyEngel** [comment]
I have a pull request opened on our internal project repository. I'll update this issue once the snapshot is ready to test 🥳

**Comment by mikasR** [comment]
OK, great. I'll wait.

**Comment by CodyEngel** [comment]
Hey @mikasR sorry for the delay, the snapshot is all set for you if you would like to test it out and this will be included in our next release.

Some things to keep in mind with our snapshots before adding them: they are not intended for use in production and will cease to work 3 months after their publish date, so please only make use of these for testing purposes.

In order to add the snapshot you will want to add an additional maven repository:
```groovy
maven {
    url 'https://oss.sonatype.org/content/repositories/snapshots/'
}
```

From there you can update the version of the Checkout SDK to be `0.2.0-SNAPSHOT` and that dependency should resolve correctly.

We also added 3DS between our 0.1.0 release and this snapshot, in order to run properly you will need to add another maven repository in order to retrieve the Cardinal SDK dependency:

```groovy
maven {
    url "https://cardinalcommerce.bintray.com/android"
    credentials {
        username 'braintree-team-sdk@cardinalcommerce'
        password '220cc9476025679c4e5c843666c27d97cfb0f951'
    }
}
```

Unfortunately this repository is a private repository and requires credentials, the above credentials are considered non-private so feel free to make use of them.

**Comment by mikasR** [comment]
OK.

I was able to test it.

The crash is gone.

The only thing i noticed, is that the onError event get called. I don't know if that's supposed to happen.
Other than that it's ok.

Tell me if i need to mantain the Cardinal SDK dependency for when this update gets released.

Thanks.«

**Comment by CodyEngel** [comment]
Hey @mikasR -- the `onError` callback can be invoked if there was an internal error with the SDK, it should provide an `ErrorInfo` instance with additional details.

As for the Cardinal SDK this dependency will be required going forward so once `0.2.0` is released it will be necessary to use the SDK. We are discussing how to make it optional in the future but I don't have an ETA on when that will be.

**Comment by mikasR** [comment]
Hi @CodyEngel , this is the content of the ErrorInfo

Error: ErrorInfo(error=java.lang.IllegalArgumentException: Return URL is null, reason=Error approving order, correlationIds=CorrelationIds(eligibilityDebugID=null, fundingEligibilityDebugID=null, updateClientConfigDebugID=4d8de8a949bbd, lsatUpgradeDebugID=null, fetchPayloadDebugID=null, currencyConversionDebugID=null, finishCheckoutDebugID=null), orderId=null, nativeSdkVersion=1.9.3-SNAPSHOT-2021-05-05T17:46+0000)

Then there's this non-fatal exception:

THIRD PARTY AUTH FAILURE Consent denied
    AuthorizationException: {"type":1,"code":1002,"error":"access_denied","errorDescription":"Consent denied","errorUri":"pt.zonesoft.biip:\/\/paypalpay"}
        at com.paypal.authcore.authentication.TokenService.getAccessToken(Unknown Source:10)
        at com.paypal.authcore.authentication.TokenService.convertCodeToToken(Unknown Source:5)
        at com.paypal.authcore.authentication.TokenActivity.onCreate(Unknown Source:12)

**Comment by mikasR** [comment]
OK, great.

Thank you.

**Comment by walelgnw** [comment]
I tap cancel and the screen goes black, and the app restarts.

I am using a paymentbutton.PayPalButton.
All i did was setup the button. Tap on it, and on the paypal webview i tap cancel and app crushes with below error

 Unable to pause activity {com.vintechplc.zoe_store.debug/com.paypal.pyplcheckout.home.view.activities.PYPLHomeActivity}: java.lang.IllegalArgumentException: NetworkCallback was already unregistered
        at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4062)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4017)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3969)
        at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:45)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1836)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6702)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:911)
     Caused by: java.lang.IllegalArgumentException: NetworkCallback was already unregistered
